### PR TITLE
clarify function name, add `find_dependencies`, `find_direct_dependencies`

### PR DESCRIPTION
**Pull request by ericphanson**
* Reading the name and docstring of `find_downstream_dependencies`, I thought it was recursive, finding all downstream dependencies of a given function, but realized it was not. Here, I changed the name to `find_direct_downstream_dependencies` which I thought was more clear, and leaves room for adding a recursive version.
  * I bumped the version number for a breaking release since this is a breaking change
* I factored out some of the code from `find_downstream_dependencies` into a separate function called `find_direct_dependencies`
* I recursively used `find_direct_dependencies` to make a new function `find_dependencies`, which finds all the direct and transitive dependencies of a given function
  * I initially called these `find_direct_upstream_dependencies` and `find_upstream_dependencies`, but I realized an "upstream dependency" is what is usually just called a "dependency". I.e. packages your package depends on
* I changed some of the UUIDs in the tests so they were all unique
* I extended the README
* I added a "Case5" test case to check that `find_dependencies` really recurses

Let me know if only some of these changes are desirable and I can factor things out. I ended up doing these all together because I was trying to understand the code and add functionality at the same time. Also, I tried to match the code style but may have missed some things.

### CODE REVIEW

**Comment by ericphanson** [comment]
By the way, my motivation is https://github.com/JuliaEcosystem/PackageAnalyzer.jl/issues/49, i.e. I want `find_dependencies` to exist so that I can use it with PackageAnalyzer to analyze package dependency trees.

For example,
```julia
using PackageAnalyzer, Pkg
using PkgDeps: PkgDeps
const stdlib_names = Set(values(Pkg.Types.stdlibs()))

function find_dependencies(pkg)
    non_stdlib_deps = filter!(!in(stdlib_names), PkgDeps.find_dependencies(pkg))
    return find_packages(non_stdlib_deps)
end

results = analyze(find_dependencies("DataFrames"))

count_julia_src(p::PackageAnalyzer.Package) = count_julia_src(p.lines_of_code)
count_julia_src(table) = sum(row.code for row in table if row.directory == "src" && row.language == :Julia; init=0)

sum(count_julia_src, results) # 27410
```

File: src/PkgDeps.jl [hunk 1]
```diff
@@ -5,7 +5,7 @@ using TOML: parsefile
 using UUIDs
 
 export PkgEntry, RegistryInstance
-export find_downstream_dependencies, reachable_registries
+export find_direct_downstream_dependencies, find_dependencies, find_direct_dependencies, reachable_registries
```
Comment (mattBrzezinski) [inline src/PkgDeps.jl]:
```suggestion
export find_dependencies, find_direct_dependencies, find_direct_downstream_dependencies, reachable_registries
```

File: src/PkgDeps.jl [hunk 2]
```diff
@@ -82,51 +82,138 @@ reachable_registries(registry_name::String; depots::Union{String, Vector{String}
...
+        latest_version = _get_latest_version(base_path)
+
+        for version_range in dependency_versions
+            if in(latest_version, VersionRange(version_range))
+                dependencies = collect(keys(deps_content[version_range]))
+                append!(latest_dependencies, dependencies)
+            end
+        end
+    end
+    return latest_dependencies
```
Comment (mattBrzezinski) [inline src/PkgDeps.jl:226]:
```suggestion

    return latest_dependencies
```

File: src/PkgDeps.jl [hunk 3]
```diff
@@ -82,51 +82,138 @@ reachable_registries(registry_name::String; depots::Union{String, Vector{String}
...
+"""
+function find_dependencies(
+    pkg_name::AbstractString;
+    registries::Array{RegistryInstance}=reachable_registries()
+)
+    return find_dependencies!(Set{String}(), pkg_name; registries)
+end
+
+# recursive helper that accumulates into `upstream_dependencies`
+function find_dependencies!(
```
Comment (mattBrzezinski) [inline src/PkgDeps.jl]: Maybe prefacing this with an underscore to denote that it's an "internal function"?

**Comment by mattBrzezinski** [comment]
Hmm I rebased, but seems to have missed a spot.

**Comment by ericphanson** [comment]
Replaced by https://github.com/JuliaEcosystem/PkgDeps.jl/pull/28. Hoping the sequence of smaller PRs is more mergable than one big one.

**Pull request closed without merging**